File: code/classes/checking-type.ex
# Checking type in Elixir (using structs and pattern matching)
defmodule Movie do
  defstruct [:name, :director]
end

defmodule Song do
  defstruct [:name, :artist]
end

library = [
  %Movie{name: "Inception", director: "Nolan"},
  %Song{name: "Bohemian Rhapsody", artist: "Queen"},
  %Movie{name: "Interstellar", director: "Nolan"}
]

{movie_count, song_count} =
  Enum.reduce(library, {0, 0}, fn
    %Movie{}, {m, s} -&gt; {m + 1, s}
    %Song{},  {m, s} -&gt; {m, s + 1}
    _,        acc    -&gt; acc
  end)

IO.puts "Movies: #{movie_count}, Songs: #{song_count}"

File: code/classes/checking-type.erl
-module(checking_type).
-export([start/0]).

start() -&gt;
    Library = [
        {movie, "Inception", "Nolan"},
        {song, "Bohemian Rhapsody", "Queen"},
        {movie, "Interstellar", "Nolan"}
    ],
    {MovieCount, SongCount} = count_types(Library, 0, 0),
    io:format("Movies: ~p, Songs: ~p~n", [MovieCount, SongCount]).

count_types([], M, S) -&gt; {M, S};
count_types([{movie, _, _} | Rest], M, S) -&gt; count_types(Rest, M + 1, S);
count_types([{song, _, _}  | Rest], M, S) -&gt; count_types(Rest, M, S + 1);
count_types([_ | Rest], M, S) -&gt; count_types(Rest, M, S).
